### PR TITLE
Simplify macros for exposing c10 ops to c2 

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17743 Improve caffe2 operator wrapping&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14362434/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#17781 Simplify macros for exposing c10 ops to c2 **&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14375054/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #17827 Expose c10 cuda ops to caffe2&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14394879/)

The wrapper for calling a c10 operator from caffe2 is now based on a runtime FunctionSchema instead of compile time information. This way, it can be created for any c10 operator schema with just one invocation to a simple macro instead of having to define arguments and more as compile time structures.

Furthermore, previously, the wrapper assumed there's an argument present for preallocated outputs, but that was only true for caffe2 operators exported to c10. So the wrapper only worked correctly for calling caffe2->c10->caffe2. Now with the new implementation, it works for any c10 operator.

Also, binary size for this should be much smaller.

Differential Revision: [D14375054](https://our.internmc.facebook.com/intern/diff/D14375054/)

### CODE REVIEW

File: caffe2/core/operator_c10wrapper.h [hunk 1]
```diff
@@ -146,31 +133,86 @@ class C10OperatorWrapper final : public Operator<Context> {
...
+      AT_ERROR(
+          "Error in caffe2->c10 wrapper: Unsupported argument type ",
+          argument.type()->str(),
+          " in c10 operator schema");
+    }
+  }
+
   c10::OperatorHandle op_;
   c10::optional<OpKernel> kernel_;
+  bool has_preallocated_outputs_;
```
Comment (ezyang) [inline caffe2/core/operator_c10wrapper.h]: This needs an explanation

File: caffe2/core/operator_c10wrapper.h [hunk 2]
```diff
@@ -25,104 +16,100 @@ using extract_type_t = typename ParameterDef::type;
...
+      Workspace* ws)
       : Operator<Context>(operator_def, ws),
-        op_(OperatorHandle()),
+        op_(op),
         kernel_(at::nullopt),
-        parameters_(parse_parameters_(
-            operator_def,
-            c10::guts::make_index_sequence<num_parameters()>())) {
+        has_preallocated_outputs_(
+            op_.schema().arguments().back().name() ==
```
Comment (ezyang) [inline caffe2/core/operator_c10wrapper.h]: So... invariant, you cannot have a C10OperatorWrapper for a schema with no arguments? (Aren't there some Caffe2 operators which don't have arguments?)

File: caffe2/core/operator_c10wrapper.h [hunk 3]
```diff
@@ -25,104 +16,100 @@ using extract_type_t = typename ParameterDef::type;
...
-    return op_.schema().arguments().size() - 1 - num_parameters();
-  }
-
-  static constexpr size_t num_parameters() {
-    return std::tuple_size<ParameterDefTuple>::value;
+    AT_ASSERT(
+        operator_def.input_size() + (has_preallocated_outputs_ ? 1 : 0) <=
+        op_.schema()
+            .arguments()
+            .size()); // '<=' because there might be caffe2 arguments
```
Comment (ezyang) [inline caffe2/core/operator_c10wrapper.h]: Would it be accurate to say that any of the trailing arguments MUST NOT be tensors?

File: caffe2/core/operator_c10wrapper.h (hunk 2 above)
Comment (smessmer) [reply caffe2/core/operator_c10wrapper.h]: good catch

File: caffe2/core/operator_c10wrapper.h (hunk 3 above)
Comment (smessmer) [reply caffe2/core/operator_c10wrapper.h]: added it to the comment

File: caffe2/core/operator_c10wrapper.h [hunk 4]
```diff
@@ -25,104 +16,100 @@ using extract_type_t = typename ParameterDef::type;
...
-    return Parameter::parse(ArgumentHelper(operator_def));
-  }
-
   void pushInputs_() {
-    if (use_array_input) {
-      stack_.emplace_back(ivalue::TensorList::create(array_inputs_()));
-    } else {
-      for (size_t i = 0; i < num_inputs(); ++i) {
-        stack_.emplace_back(at::Tensor(C10Tensor(Input(i))));
+    AT_ASSERT(stack_.size() == 0);
```
Comment (ezyang) [inline caffe2/core/operator_c10wrapper.h]: Umm, why? The whole point of a stack machine is that you don't have to allocate a fresh argument vector every time you make a call; you can use the suffix to hold whatever set of arguments you need for the op you're running (but the prefix gets ignored.)

File: caffe2/core/operator_c10wrapper.h [hunk 5]
```diff
@@ -25,104 +16,100 @@ using extract_type_t = typename ParameterDef::type;
...
+            input_tensor_index < InputSize(),
+            "Error in caffe2->c10 wrapper: Too few tensor arguments given (",
+            InputSize(),
+            "), operator schema expected more.");
+        stack_.emplace_back(at::Tensor(Input(input_tensor_index++)));
+
+      } else if (argument.type()->isSubtypeOf(ListType::ofTensors())) {
+        AT_ASSERTM(
+            input_tensor_index == 0,
+            "Error in caffe2->c10 wrapper: Schema can only have either one or more Tensor inputs or one TensorList input.");
```
Comment (ezyang) [inline caffe2/core/operator_c10wrapper.h]: You're not actually checking this invariant, though, if the argument order is `TensorList, Tensor`, is that right?

File: caffe2/core/operator_c10wrapper.h [hunk 6]
```diff
@@ -146,31 +133,86 @@ class C10OperatorWrapper final : public Operator<Context> {
...
+  std::vector<at::Tensor> preallocated_outputs_() {
+    std::vector<at::Tensor> result;
+    result.reserve(OutputSize());
+    for (size_t i = 0; i < OutputSize(); ++i) {
+      result.emplace_back(OperatorBase::OutputTensorOrUndefined(i));
     }
     return result;
   }
 
+  IValue get_nontensor_argument_(const c10::Argument& argument) {
```
Comment (ezyang) [inline caffe2/core/operator_c10wrapper.h]: This method looks very copy pastey. OK for now but worth seeing if it can be simplified.

**Review by ezyang — APPROVED** [review]

File: caffe2/core/operator_c10wrapper.h (hunk 5 above)
Comment (smessmer) [reply caffe2/core/operator_c10wrapper.h]: This case sets `input_tensor_index = InputSize()` at line 100, which then fails the assertion in line 88 if an argument of type `Tensor` follows.

**Pull request closed without merging**